File: internal/cli/prepare.go
package cli

import (
	"context"
	"crypto/sha256"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os"
	"strings"
	"time"

	"github.com/atyrode/babel/internal/adapter"
	"github.com/atyrode/babel/internal/complaint"
	"github.com/atyrode/babel/internal/digest"
	"github.com/atyrode/babel/internal/event"
	"github.com/atyrode/babel/internal/frontier"
	"github.com/atyrode/babel/internal/index"
	runstore "github.com/atyrode/babel/internal/run"
)

const prepareUsage = `Usage: babel prepare [SELECTOR...] [flags]

Fixes an exploration's corpus scope and emits an immutable preparation ID
(SPEC.md §6.5, §8). With no selector the scope is every session the source
adapters can see on this host; with selectors it is exactly those sessions.

The identity is derived from the content of the selection — host, harness,
source identity, and both digests per session — so naming it on a later
"babel explore --preparation ID" states which corpus that run read instead
of leaving it to whatever the machine happened to hold at the time.

Each selected session is also read into the local retrieval index, which is
the corpus a run's search tool is served from. The index is a rebuildable
cache under the cache directory; the preparation record is durable state.

Babel's own prior output is indexed alongside it and searched with the
scope's own salient terms, mechanically and with no model involved. The
records that come back are recorded in the preparation, and the run that
explores it receives them as prior candidate ideas to refine, revive, or
amend rather than duplicate (issue #87).

Nothing is sent anywhere and no repository is opened: this command reads
local files and writes local state only.

Flags:
  --harness NAME       restrict to one harness: omp, codex, or claude
  --roots DIR[,DIR]    scan these roots instead of the adapter defaults
  --host ID            archive host identity recorded in the selection
                       (default $BABEL_HOST_ID, else storage.json, else this
                       machine's hostname)
  --serendipitous      mark the scope as drawn for exploration, so the
                       related prior outputs reach the run as inspiration
                       rather than as constraint
  --json               emit the preparation as JSON on stdout

A selector is "HARNESS/SOURCE-ID", or any unambiguous suffix of one.
`

// relatedRow is one prior Babel output a preparation names, in the
// machine-readable document. The summary is not stored in the preparation —
// only the kind and the id are — so what is shown here is read back from the
// frontier and is the record's current wording.
type relatedRow struct {
	Kind    string  `json:"kind"`
	ID      string  `json:"id"`
	Summary string  `json:"summary"`
	Status  string  `json:"status,omitempty"`
	Overlap float64 `json:"-"`
}

// preparedRow is one session in a preparation's machine-readable document.
// Digests are never truncated: a caller has to be able to compare them.
type preparedRow struct {
	Harness       string `json:"harness"`
	SourceID      string `json:"source_id"`
	Selector      string `json:"selector"`
	Host          string `json:"host"`
	CaptureDigest string `json:"capture_digest"`
	SourceDigest  string `json:"source_digest"`
	Bytes         int64  `json:"bytes"`
	Records       int    `json:"records"`
	Events        int    `json:"events"`
}

// prepareResult is `babel prepare --json`.
type prepareResult struct {
	PreparationID string        `json:"preparation_id"`
	PreparedAt    string        `json:"prepared_at"`
	Host          string        `json:"host"`
	Sessions      []preparedRow `json:"sessions"`
	IndexedEvents int           `json:"indexed_events"`
	// FrontierRecords is how many of Babel's own outputs the index now
	// holds, and SalientTerms is the query the scope produced. Both are
	// reported because the injection is mechanical: an operator who sees
	// unexpected related records needs to be able to see the terms that
	// found them.
	FrontierRecords int          `json:"frontier_records"`
	SalientTerms    []string     `json:"salient_terms,omitempty"`
	Related         []relatedRow `json:"related,omitempty"`
	Serendipitous   bool         `json:"serendipitous,omitempty"`
	Database        string       `json:"database"`
	Index           string       `json:"index"`
}

// prepare implements `babel prepare`.
func (a *app) prepare(ctx context.Context, args []string) error {
	c := newCmd("prepare", prepareUsage)
	var sf scanFlags
	var rf repoFlags
	sf.bindHarness(c)
	sf.bindRoots(c)
	c.fs.StringVar(&rf.host, "host", "", "archive host identity recorded in the selection")
	serendipitous := c.fs.Bool("serendipitous", false,
		"mark the scope as drawn for exploration rather than to answer something")
	asJSON := c.fs.Bool("json", false, "emit the preparation as JSON")
	if err := c.parse(a, args); err != nil {
		return err
	}
	ads, err := sf.selected(c)
	if err != nil {
		return err
	}
	host, err := rf.hostID(c)
	if err != nil {
		return err
	}

	sessions, _ := a.scan(ctx, ads, sf.rootList())
	chosen, err := selectSessions(c, sessions, c.args())
	if err != nil {
		return err
	}

	d, err := babelDirs()
	if err != nil {
		return err
	}
	runs, err := runstore.Open(d.durableDir())
	if err != nil {
		return err
	}
	defer runs.Close()

	scoped, err := a.fixScope(ctx, runs, chosen, host, *serendipitous)
	if err != nil {
		return err
	}

	res := prepareResult{
		PreparationID:   string(scoped.prep.ID),
		PreparedAt:      formatTime(scoped.prep.PreparedAt),
		Host:            Sanitize(host),
		Sessions:        scoped.rows,
		IndexedEvents:   scoped.indexed,
		FrontierRecords: scoped.frontierRecords,
		SalientTerms:    sanitizeAll(scoped.terms),
		Related:         scoped.related,
		Serendipitous:   scoped.prep.Serendipitous,
		Database:        Sanitize(runs.Path()),
		Index:           Sanitize(scoped.index),
	}
	if *asJSON {
		return a.emitJSON(res)
	}
	table := make([][]string, 0, len(scoped.rows))
	for _, row := range scoped.rows {
		table = append(table, []string{row.Selector, fmt.Sprintf("%d", row.Bytes), fmt.Sprintf("%d", row.Events)})
	}
	if err := writeTable(a.stdout, []string{"SELECTOR", "BYTES", "EVENTS"}, table); err != nil {
		return err
	}
	fmt.Fprintf(a.stdout, "\npreparation %s over %d %s\n", res.PreparationID,
		len(scoped.rows), plural(len(scoped.rows), "session", "sessions"))
	fmt.Fprintf(a.stdout, "explore it with: babel explore --preparation %s\n", res.PreparationID)
	if len(res.Related) > 0 {
		framing := "to refine, revive, or amend rather than duplicate"
		if res.Serendipitous {
			framing = "as inspiration, not constraint"
		}
		fmt.Fprintf(a.stdout, "\n%d related prior %s, %s:\n",
			len(res.Related), plural(len(res.Related), "output", "outputs"), framing)
		for _, row := range res.Related {
			fmt.Fprintf(a.stdout, "  %s %s  %s\n", row.Kind, row.ID, row.Summary)
		}
	}
	return nil
}

// scopedCorpus is what fixing one exploration's corpus scope produced: the
// immutable preparation record, the per-session rows a caller reports, how many
// events reached the retrieval index, and where that index lives.
type scopedCorpus struct {
	prep            runstore.Preparation
	rows            []preparedRow
	indexed         int
	frontierRecords int
	terms           []string
	related         []relatedRow
	index           string
}

// fixScope digests and indexes the chosen sessions, derives the preparation
// identity from their content, and records it.
//
// It is the whole of §6.5's preparation step, shared by `babel prepare` and by a
// conductor cycle. Sharing it is not tidiness: a preparation's identity is
// derived from what was selected and digested, so a second implementation that
// digested or normalized differently would mint a different identity for the
// same corpus and make two runs over one scope look like runs over two.
//
// Storing is part of it. A preparation is content-addressed and idempotent
// (§8), so a repeated scope is the same record rather than a second one, and a
// caller that fixed a scope without recording it would hold an identity nothing
// could later resolve.
func (a *app) fixScope(ctx context.Context, runs *runstore.Store,
	chosen []localSession, host string, serendipitous bool) (scopedCorpus, error) {
	d, err := babelDirs()
	if err != nil {
		return scopedCorpus{}, err
	}
	idx, err := index.Open(d.indexDir())
	if err != nil {
		return scopedCorpus{}, err
	}
	defer idx.Close()

	version := readBuildIdentity().Version
	out := scopedCorpus{
		rows:  make([]preparedRow, 0, len(chosen)),
		index: idx.Path(),
	}
	selection := make([]runstore.Selected, 0, len(chosen))
	// The scope's salient terms are accumulated during the digest pass, which
	// already reads every event of every selected session. Deriving them
	// there rather than in a third pass is what keeps the frontier query free
	// on a corpus whose only real cost is bytes read.
	salience := index.NewSalience()
	for n, s := range chosen {
		// A cold preparation reads every selected session twice, once to
		// digest and once to index, which takes minutes on a large corpus:
		// it narrates on stderr so the wait is never silent, while stdout
		// keeps carrying exactly one document.
		a.diagf("preparing %d/%d %s...\n", n+1, len(chosen), Sanitize(s.key()))
		desc, err := describe(ctx, s)
		if err != nil {
			return scopedCorpus{}, err
		}
		stream := event.Stream{
			Harness:       s.src.Harness,
			AdapterSchema: s.owner.Schema(),
			SourceID:      s.src.SourceID,
			Path:          s.src.PrimaryPath,
		}
		capture, source, size, err := streamDigests(stream, salience)
		if err != nil {
			return scopedCorpus{}, err
		}
		result, err := idx.IndexSession(ctx, stream)
		if err != nil {
			return scopedCorpus{}, fmt.Errorf("index %s: %w", s.key(), err)
		}
		out.indexed += result.Events
		selection = append(selection, runstore.Selected{
			Host:          host,
			Harness:       s.src.Harness,
			SourceID:      s.src.SourceID,
			CaptureDigest: capture,
			SourceDigest:  source,
			Adapter: runstore.AdapterRef{
				Schema:       s.owner.Schema(),
				Version:      version,
				Completeness: completenessOf(desc),
			},
		})
		out.rows = append(out.rows, preparedRow{
			Harness:       Sanitize(s.src.Harness),
			SourceID:      Sanitize(s.src.SourceID),
			Selector:      Sanitize(s.key()),
			Host:          Sanitize(host),
			CaptureDigest: string(capture),
			SourceDigest:  string(source),
			Bytes:         size,
			Records:       result.Records,
			Events:        result.Events,
		})
	}

	// The frontier's own retrieval surface is reconciled here, with the
	// sessions, because this is the command that fixes a scope: the related
	// outputs a preparation records have to be the ones the frontier held
	// when it was fixed, not the ones it held whenever some earlier command
	// last looked.
	front, err := frontier.Open(d.durableDir())
	if err != nil {
		return scopedCorpus{}, err
	}
	defer front.Close()
	// The operator's complaints are reconciled in the same pass, which is what
	// makes them steering rather than a suggestion box (#115): a preparation
	// that surfaced Babel's own prior output and not the operator's standing
	// annoyances would relate this scope to everything except the thing a
	// person actually said about it.
	told, err := complaint.Open(d.durableDir())
	if err != nil {
		return scopedCorpus{}, err
	}
	defer told.Close()
	terms := salience.Terms(0)
	frontierRecords, related, err := a.relatedOutputs(ctx, idx, front, told, terms)
	if err != nil {
		return scopedCorpus{}, err
	}
	refs := make([]runstore.RelatedOutput, 0, len(related))
	for _, row := range related {
		refs = append(refs, runstore.RelatedOutput{Kind: row.Kind, ID: row.ID})
	}
	out.frontierRecords = frontierRecords
	out.terms = terms
	out.related = related

	prep, err := runstore.NewPreparation(time.Now().UTC(), selection, runstore.PreparationContext{
		Related:       refs,
		Serendipitous: serendipitous,
	})
	if err != nil {
		return scopedCorpus{}, fmt.Errorf("fix the corpus scope: %w", err)
	}
	if err := runs.PutPreparation(ctx, prep); err != nil {
		return scopedCorpus{}, fmt.Errorf("record preparation %s: %w", prep.ID, err)
	}
	out.prep = prep
	return out, nil
}

// selectSessions narrows a scan to the named selectors, or to everything
// when none were given. A selector that matches nothing is a rejected
// invocation rather than a silently smaller scope, because a preparation
// records what an operator meant to explore.
func selectSessions(c *cmd, sessions []localSession, selectors []string) ([]localSession, error) {
	if len(selectors) == 0 {
		if len(sessions) == 0 {
			return nil, fmt.Errorf("no local sessions to prepare")
		}
		return sessions, nil
	}
	var chosen []localSession
	seen := make(map[string]struct{}, len(selectors))
	for _, selector := range selectors {
		s, err := resolveSelector(c, sessions, selector)
		if err != nil {
			return nil, err
		}
		if _, dup := seen[s.key()]; dup {
			continue
		}
		seen[s.key()] = struct{}{}
		chosen = append(chosen, s)
	}
	return chosen, nil
}

// relatedOutputs reconciles the frontier surface of the retrieval index and
// asks it, with the scope's own salient terms, what Babel has already said
// about material like this (#87 item 4).
//
// It is mechanical from end to end: the terms come from term frequency against
// document frequency over the prepared sessions, the query is the same
// optional-term FTS expression a worker's search would use, and the page is
// bounded by run.MaxRelatedOutputs. No model is consulted and none could be —
// this runs inside `babel prepare`, which sends nothing anywhere.
//
// A scope with nothing searchable in it, or a frontier holding nothing, yields
// no related records and no error. That is a real answer: the first
// preparation on a machine has no prior output to relate to, and a preparation
// that pretended otherwise would be recording an empty gesture in an immutable
// record.
func (a *app) relatedOutputs(ctx context.Context, idx *index.Index, front *frontier.Store,
	told *complaint.Store, terms []string) (int, []relatedRow, error) {
	outputs, err := front.Outputs(ctx)
	if err != nil {
		return 0, nil, fmt.Errorf("read the frontier: %w", err)
	}
	// One reconcile over both sources, because IndexFrontier deletes the local
	// rows the set does not name: a frontier-only pass here would delete every
	// complaint capture indexed, and the next `babel tell` would put them back.
	outputs, err = complaint.Append(ctx, told, outputs)
	if err != nil {
		return 0, nil, err
	}
	indexed, err := idx.IndexFrontier(ctx, outputs)
	if err != nil {
		return 0, nil, fmt.Errorf("index the frontier: %w", err)
	}
	if len(terms) == 0 || indexed.Records == 0 {
		return indexed.Records, nil, nil
	}
	hits, err := idx.FrontierSearch(ctx, index.FrontierQuery{
		Match: strings.Join(terms, " "),
		Order: index.OrderRelevance,
		Limit: runstore.MaxRelatedOutputs,
	})
	if err != nil {
		// An unsearchable query is not a failed preparation. The scope
		// produced no term the tokenizer could match, which is a fact about
		// the scope, and the preparation records no related outputs.
		if errors.Is(err, index.ErrNoSearchableTerm) || errors.Is(err, index.ErrMatchTooLong) {
			return indexed.Records, nil, nil
		}
		return 0, nil, fmt.Errorf("search the frontier: %w", err)
	}
	related := make([]relatedRow, 0, len(hits))
	for _, hit := range hits {
		related = append(related, relatedRow{
			Kind:    Sanitize(string(hit.Kind)),
			ID:      Sanitize(hit.ID),
			Summary: Sanitize(hit.Summary),
			Status:  Sanitize(string(hit.Status)),
		})
	}
	return indexed.Records, related, nil
}

// streamDigests reads one primary log once and returns both digests §7
// requires of a selection entry, plus the bytes read. It also feeds the
// scope's salience accumulator, when one is supplied.
//
// The capture digest covers the file as it is on disk, which is what a
// restore is checked against. The source digest covers the normalized event
// stream, which is what analysis actually reads: a normalization change
// moves the second without moving the first, and that difference is what a
// later reviewer needs in order to tell "the corpus changed" from "our
// reading of it changed". Both come from one pass, because the corpus is
// dominated by a handful of very large sessions and reading them twice per
// preparation would double the only cost that matters — which is also why
// the salient terms are counted here instead of in a pass of their own.
func streamDigests(stream event.Stream, salience *index.Salience) (capture, source digest.Digest, size int64, err error) {
	file, err := os.Open(stream.Path)
	if err != nil {
		return "", "", 0, fmt.Errorf("open %s: %w", stream.Path, err)
	}
	defer file.Close()

	captureHash := sha256.New()
	sourceHash := sha256.New()
	encoder := json.NewEncoder(sourceHash)
	counted := &countingReader{r: io.TeeReader(file, captureHash)}
	if err := event.Scan(counted, stream, func(e event.Event) error {
		if salience != nil {
			salience.Add(e.Text)
		}
		return encoder.Encode(e)
	}); err != nil {
		return "", "", 0, fmt.Errorf("read %s: %w", stream.Path, err)
	}
	// Whatever the scanner did not consume still belongs to the capture: the
	// capture digest identifies the file, not the part of it that parsed.
	if _, err := io.Copy(io.Discard, counted); err != nil {
		return "", "", 0, fmt.Errorf("read %s: %w", stream.Path, err)
	}
	return digest.New([sha256.Size]byte(captureHash.Sum(nil))),
		digest.New([sha256.Size]byte(sourceHash.Sum(nil))),
		counted.n, nil
}

// countingReader counts the bytes read through it, which is how the capture
// size is learned without a second stat that could disagree with what was
// hashed.
type countingReader struct {
	r io.Reader
	n int64
}

func (c *countingReader) Read(p []byte) (int, error) {
	n, err := c.r.Read(p)
	c.n += int64(n)
	return n, err
}

// completenessOf is the adapter's account of the metadata it could not
// derive. It is copied into the preparation because §3 forbids synthesizing
// a value to satisfy a shape, and a scope record that hid the gaps would be
// claiming a fidelity the adapter did not report.
func completenessOf(desc *adapter.Description) []adapter.CompletenessReason {
	if desc == nil {
		return nil
	}
	return desc.Meta.Completeness
}

File: plugins/atyrode.babel/machine/prepare.ts
/*
  THE prepare OPERATION (plan §4): selectors in, a preparation out — an immutable statement of
  one exploration's corpus scope, whose identity is DERIVED from the content of the selection
  rather than assigned.

  That is what makes `explore --preparation <id>` mean something: naming a preparation states
  which corpus a run read, instead of leaving it to whatever the machine happened to hold at
  the time. Two preparations over the same sessions at the same captures are the same id, so
  two runs over one scope look like two runs over one scope; one session's bytes changing is a
  different id, because it is a different corpus.

  DELIBERATE DIFFERENCE FROM THE GO PRODUCT (internal/run/preparation.go): `preparedAt` is
  recorded but NOT hashed. Go's derivation included the instant, so re-preparing an unchanged
  corpus minted a second identity for the same scope — which contradicts the idempotence that
  same-scope-same-id is for. Here the id is a function of the selection alone.

  Each entry carries both digests SPEC §7 requires of a selection: the CAPTURE digest over the
  primary log's bytes as they lie on disk, which is what a restore is checked against, and the
  SOURCE digest over the normalized record stream, which is what analysis reads. A harness that
  rewrites its log with different spacing moves the first and not the second, and that
  difference is what a later reviewer needs in order to tell "the corpus changed" from "our
  reading of it changed".

  What the normalization IS, in this wave: one canonical JSON record per line — object keys
  ordered, insignificant whitespace gone — and an explicit opaque marker for a line that is not
  a record, so nothing is ever dropped. What it is NOT, yet: internal/event's classification of
  each record into §6.3's five evidence kinds, which belongs with the retrieval index that is
  its only consumer. When that lands it owns the source digest and bumps PREPARATION_SCHEMA.
*/

import { z } from "zod";
import type { Receipt } from "../contract.ts";
import type { SessionRef } from "./adapters/index.ts";
import type { OutputSink } from "./output.ts";

export const PrepareInputSchema = z.strictObject({
  /** The run this job is; empty mints one (see archive.ts). */
  runId: z.string().trim().max(120).default(""),
  /** The machine's identity, recorded as the host of every session it holds itself. */
  machineId: z.string().trim().min(1).max(120),
  /** `HARNESS/SOURCE-ID`, or any unambiguous suffix of one. Empty scopes every session this
   *  machine can see, which is what a scheduled preparation wants. */
  selectors: z.array(z.string().trim().min(1).max(400)).max(500).default([]),
});
export type PrepareInput = z.infer<typeof PrepareInputSchema>;

export interface SessionDigests {
  readonly captureDigest: string;
  readonly sourceDigest: string;
  readonly bytes: number;
}

/** The machine facts this operation needs, which the adapters own (machine/adapters). */
export interface PrepareDeps {
  /** Every session this machine can see, under the adapters' own roots. */
  discover(): Promise<readonly SessionRef[]>;
  digests(ref: SessionRef): Promise<SessionDigests>;
}

/** The version of the preparation record's shape AND of the normalization behind its source
 *  digest. It participates in the derivation, so a record written by a later schema can never
 *  collide with one written by this schema even if every other field matches. */
export const PREPARATION_SCHEMA = 2;

/** Separates this hash from every other use of SHA-256 in Babel: without a domain, a digest
 *  over some other structure that happened to serialize identically would be a valid
 *  preparation id. The `v2` is PREPARATION_SCHEMA's own, moved with it. */
const PREPARATION_DOMAIN = "babel/preparation/v2";

/** Marks a preparation id as one, so a mistyped identifier fails as the wrong kind of id
 *  rather than as a missing row. */
const PREPARATION_PREFIX = "prep-";

/** One session inside a preparation, identified as decision 9 identifies sessions: the machine
 *  that holds it, the harness, and the adapter-defined source identity. */
export type PreparationEntry = {
  readonly host: string;
  readonly harness: string;
  readonly sourceId: string;
  readonly captureDigest: string;
  readonly sourceDigest: string;
};

export type Preparation = {
  readonly id: string;
  readonly schema: number;
  /** When the scope was fixed. Recorded, never hashed — see the header. */
  readonly preparedAt: string;
  /** Canonically ordered, so neither discovery order nor a caller's later mutation can change
   *  what the record means. */
  readonly selection: readonly PreparationEntry[];
};

/** One `sessions` row as a preparation observed it: the identity, the capture it fixed, and
 *  the size that capture was. Nothing else — `scan` owns the rest of the row. */
interface PreparedSessionRow {
  readonly selector: string;
  readonly host: string;
  readonly harness: string;
  readonly source_id: string;
  readonly content_digest: string;
  readonly size: number;
  readonly seen_at: string;
}

/** A record longer than this is hashed in bounded pieces, split at a content-determined
 *  offset so the digest never depends on how the file happened to arrive in chunks. */
const MAX_RECORD_CHARS = 4 << 20;

/**
 * Fixes a corpus scope and derives its identity.
 *
 * An empty selection is refused: an exploration over nothing is a mistake, and accepting it
 * would make a broken selection indistinguishable from a deliberate one. So is the same
 * session twice, which would let a scope claim a weight it does not have.
 */
export function newPreparation(preparedAt: string, selection: readonly PreparationEntry[]): Preparation {
  if (selection.length === 0) throw new Error("preparation: the selection is empty");
  const canonical = [...selection].sort((a, b) => {
    if (a.host !== b.host) return a.host < b.host ? -1 : 1;
    if (a.harness !== b.harness) return a.harness < b.harness ? -1 : 1;
    if (a.sourceId === b.sourceId) return 0;
    return a.sourceId < b.sourceId ? -1 : 1;
  });
  for (const [index, entry] of canonical.entries()) {
    if (entry.host === "" || entry.harness === "" || entry.sourceId === "") {
      throw new Error(`preparation: selection entry ${index} names no session`);
    }
    if (entry.captureDigest === "") {
      throw new Error(`preparation: selection entry ${index} has no capture digest`);
    }
    const previous = canonical[index - 1];
    if (
      previous !== undefined &&
      previous.host === entry.host &&
      previous.harness === entry.harness &&
      previous.sourceId === entry.sourceId
    ) {
      throw new Error(`preparation: the selection holds ${entry.harness}/${entry.sourceId} twice`);
    }
  }
  return { id: derive(canonical), schema: PREPARATION_SCHEMA, preparedAt, selection: canonical };
}

/**
 * The id, as `prep-<64 lowercase hex>`.
 *
 * The hashed encoding is injective, which is what makes the id a function of the content
 * rather than of its punctuation:
 *
 *   PREPARATION_DOMAIN || u32(schema) || u32(entries)
 *     then per entry, in canonical order:
 *       lp(host) || lp(harness) || lp(sourceId) || lp(captureDigest) || lp(sourceDigest)
 *
 * where lp(s) is a four-byte big-endian length followed by s's UTF-8 bytes and u32 is four big
 * endian bytes. Every variable-length field is length-prefixed and every repetition is
 * count-prefixed, so no two distinct selections encode to the same bytes — plain concatenation
 * would let a host of "a" with a harness of "bc" collide with a host of "ab" and a harness
 * of "c".
 */
function derive(selection: readonly PreparationEntry[]): string {
  const hasher = new Bun.CryptoHasher("sha256");
  hasher.update(PREPARATION_DOMAIN);
  writeU32(hasher, PREPARATION_SCHEMA);
  writeU32(hasher, selection.length);
  for (const entry of selection) {
    writeLP(hasher, entry.host);
    writeLP(hasher, entry.harness);
    writeLP(hasher, entry.sourceId);
    writeLP(hasher, entry.captureDigest);
    writeLP(hasher, entry.sourceDigest);
  }
  return PREPARATION_PREFIX + hasher.digest("hex");
}

/** One reusable four-byte frame for the length prefixes: `update` copies synchronously and
 *  nothing awaits between the write and the read, so one buffer serves every derivation. */
const U32_BYTES = new Uint8Array(4);
const U32 = new DataView(U32_BYTES.buffer);

function writeU32(hasher: Bun.CryptoHasher, value: number): void {
  U32.setUint32(0, value, false);
  hasher.update(U32_BYTES);
}

function writeLP(hasher: Bun.CryptoHasher, value: string): void {
  const bytes = new TextEncoder().encode(value);
  writeU32(hasher, bytes.length);
  hasher.update(bytes);
}

/**
 * Both digests of one session's primary log, and the bytes they covered, from ONE pass.
 *
 * One pass because the corpus is dominated by a handful of very large logs, and reading them
 * twice per preparation would double the only cost that matters. The capture digest covers
 * every byte, including whatever the record splitter did not find a record in: it identifies
 * the file, not the part of it that parsed.
 */
export async function digests(ref: SessionRef): Promise<SessionDigests> {
  const capture = new Bun.CryptoHasher("sha256");
  const source = new Bun.CryptoHasher("sha256");
  const decoder = new TextDecoder();
  let bytes = 0;
  let pending = "";
  const record = (line: string): void => {
    const normalized = normalize(line);
    if (normalized !== "") source.update(normalized);
  };
  for await (const chunk of Bun.file(ref.primaryPath).stream()) {
    capture.update(chunk);
    bytes += chunk.byteLength;
    pending += decoder.decode(chunk, { stream: true });
    let start = 0;
    for (let nl = pending.indexOf("\n"); nl >= 0; nl = pending.indexOf("\n", start)) {
      record(pending.slice(start, nl));
      start = nl + 1;
    }
    pending = start === 0 ? pending : pending.slice(start);
    while (pending.length > MAX_RECORD_CHARS) {
      record(pending.slice(0, MAX_RECORD_CHARS));
      pending = pending.slice(MAX_RECORD_CHARS);
    }
  }
  pending += decoder.decode();
  record(pending);
  return {
    captureDigest: `sha256:${capture.digest("hex")}`,
    sourceDigest: `sha256:${source.digest("hex")}`,
    bytes,
  };
}

/**
 * One line of a primary log as the normalized stream states it.
 *
 * A record becomes canonical JSON, so a harness that reorders its keys or reflows its
 * whitespace is the same corpus. A line that is not a record is retained verbatim behind a
 * marker canonical JSON can never start with: a torn or corrupt line is evidence that this
 * exact line was seen, and dropping it would make degradation silent.
 */
function normalize(line: string): string {
  if (line.trim() === "") return "";
  try {
    return `${canonical(JSON.parse(line))}\n`;
  } catch {
    return `!${line}\n`;
  }
}

function canonical(value: unknown): string {
  if (value === null || typeof value !== "object") return JSON.stringify(value) ?? "null";
  if (Array.isArray(value)) return `[${value.map(canonical).join(",")}]`;
  const keys = Object.keys(value).sort();
  const fields = keys.map((key) => `${JSON.stringify(key)}:${canonical((value as Record<string, unknown>)[key])}`);
  return `{${fields.join(",")}}`;
}

export async function prepare(
  input: PrepareInput,
  out: OutputSink,
  deps: PrepareDeps,
): Promise<Receipt> {
  const startedAt = new Date().toISOString();
  const runId = input.runId === "" ? `run_${crypto.randomUUID()}` : input.runId;
  const counts = { discovered: 0, selected: 0, bytes: 0 };
  const rows: PreparedSessionRow[] = [];
  let preparation: Preparation | null = null;
  let closure: Receipt["closure"] = "completed";
  let reason = "";

  const discovered = await deps.discover();
  counts.discovered = discovered.length;
  const chosen = choose(discovered, input.selectors);
  if (chosen.failure !== "") {
    // A selector that matches nothing, or matches two sessions, is a rejected invocation
    // rather than a silently smaller scope: a preparation records what was meant to be
    // explored, and a scope that quietly shrank would make the next run's coverage a mystery.
    closure = "failed";
    reason = chosen.failure;
  } else if (chosen.chosen.length === 0) {
    closure = "skipped";
    reason = "no session on this machine to prepare";
  } else {
    const seenAt = new Date().toISOString();
    const selection: PreparationEntry[] = [];
    for (const session of chosen.chosen) {
      let measured;
      try {
        measured = await deps.digests(session);
      } catch (err) {
        // The scope is refused whole. A preparation missing one of the sessions it was asked
        // for would be an immutable record of a corpus nobody chose.
        closure = "failed";
        reason = `read ${session.selector}: ${err instanceof Error ? err.message : String(err)}`;
        break;
      }
      counts.bytes += measured.bytes;
      selection.push({
        host: input.machineId,
        harness: session.harness,
        sourceId: session.sourceId,
        captureDigest: measured.captureDigest,
        sourceDigest: measured.sourceDigest,
      });
      rows.push({
        selector: session.selector,
        host: input.machineId,
        harness: session.harness,
        source_id: session.sourceId,
        content_digest: measured.captureDigest,
        size: measured.bytes,
        seen_at: seenAt,
      });
    }
    if (closure === "completed") {
      preparation = newPreparation(new Date().toISOString(), selection);
      counts.selected = selection.length;
    }
  }

  // Every scoped session is registered in the hub's catalog, because a preparation that named
  // sessions the hub holds no row for would be an identity nothing could later resolve.
  await out.write("sessions", closure === "completed" ? rows : []);
  const receipt: Receipt = {
    runId,
    kind: "prepare",
    machineId: input.machineId,
    startedAt,
    finishedAt: new Date().toISOString(),
    closure,
    counts: { ...counts },
    ...(preparation === null ? {} : { preparation }),
    ...(reason === "" ? {} : { reason }),
  };
  await out.receipt(receipt);
  return receipt;
}

/**
 * The sessions the selectors name, or the first refusal.
 *
 * Matching is tiered — the exact canonical selector first, then a segment-aligned suffix, then
 * any suffix — so a fully qualified selector can never be shadowed by a longer source id that
 * happens to end the same way. Ambiguity is a rejected invocation, not a guess, and the
 * candidates are listed so the operator can qualify what they meant.
 */
function choose(
  sessions: readonly SessionRef[],
  selectors: readonly string[],
): { readonly chosen: readonly SessionRef[]; readonly failure: string } {
  if (selectors.length === 0) return { chosen: sessions, failure: "" };
  const chosen: SessionRef[] = [];
  const taken = new Set<string>();
  for (const selector of selectors) {
    const exact: SessionRef[] = [];
    const aligned: SessionRef[] = [];
    const loose: SessionRef[] = [];
    for (const session of sessions) {
      if (session.selector === selector) exact.push(session);
      else if (session.selector.endsWith(`/${selector}`)) aligned.push(session);
      else if (session.selector.endsWith(selector)) loose.push(session);
    }
    const tier = exact.length > 0 ? exact : aligned.length > 0 ? aligned : loose;
    const [session] = tier;
    if (session === undefined) {
      return { chosen: [], failure: `no session on this machine matches the selector ${selector}` };
    }
    if (tier.length > 1) {
      return {
        chosen: [],
        failure: `the selector ${selector} is ambiguous, it matches ${tier.length} sessions: ${tier
          .map((candidate) => candidate.selector)
          .join(" ")}`,
      };
    }
    if (taken.has(session.selector)) continue;
    taken.add(session.selector);
    chosen.push(session);
  }
  return { chosen, failure: "" };
}
